---
title: "Microcystis Oligotype Analysis"
author: "Michelle Berry"
date: "December 2016"
output:
  html_document: default
  pdf_document: default
---

```{r global_options, echo = FALSE}

knitr::opts_chunk$set(
  fig.align = 'center', 
  warning = FALSE, 
  message = FALSE
)

## To run this document use the following line
# rmarkdown::render("mc-oligotype-analysis.Rmd", output_format = c("pdf_document", "html_document"))

```

```{r message=FALSE}
library(ggplot2)
library(dplyr)
library(grid)
library(gridExtra)
library(cowplot)
library(gtable)
library(tidyr)
library(pander)
library(ape)
theme_set(theme_bw())

```


# Import Data
```{r}
# Import oligotypes
mc_oligos <- read.csv(
  file = "data/mc-oligo-counts.txt", 
  sep = "\t"
)

# Remove isolate data (keeping just erie samples)
erie_oligos <- 
  mc_oligos %>%
    filter(grepl(pattern = "E0", x = samples))

sampdat <- read.csv("data/erie-sampledata.csv")



```

# Format data
```{r}
  
# Join sample data to oligotype data
oligos_join <- 
  sampdat %>%
    left_join(erie_oligos, by = c("SampleID" = "samples"))

# Scale oligotype data by total reads from samples
oligos_scale <-
  oligos_join %>%
    mutate(CTT = CTT/ReadSums) %>%
    mutate(CTG = CTG/ReadSums) %>%
    mutate(CCG = CCG/ReadSums) %>%
    mutate(CTG_CCG_ratio = CTG/CCG)

# Tidy oligotypes into one column
erie_oligo_df <- 
  oligos_scale %>%
  select(-CCT, -TCG) %>%
  gather(key = Oligotype, value = count, CTG, CTT, CCG)


# Mean number of reads per sample
mean(erie_oligo_df$ReadSums)

# Median number of reads per sample
hist(erie_oligo_df$ReadSums, main = "Histogram of sample read counts", xlab = "Read Sums")
```

# Plot data
```{r}
# Fixes the ordering of date factors
order_dates <- function(df){
  df$Date <- factor(df$Date, 
    levels = c("6/16","6/30","7/8","7/14","7/21",
      "7/29","8/4","8/11","8/18","8/25","9/2","9/8","9/15",
      "9/23","9/29","10/6","10/15","10/20","10/27"))
  return(df)
}

erie_oligo_df <- order_dates(erie_oligo_df)
erie_oligo_df$Oligotype <- factor(erie_oligo_df$Oligotype, levels = c("CCG", "CTT", "CTG"))


# Plot oligotypes over time
oligoplot <- 
  ggplot(data = erie_oligo_df, aes(x = Date, y = count, fill = Oligotype)) +
    facet_grid(~Station) +
    geom_bar(stat = "identity") +
    scale_fill_manual(values = c("#b3002d", "#43a2ca", "#a6d854")) + 
    scale_x_discrete(
      breaks = c("7/8", "8/4", "9/2", "10/6"),
      labels = c("Jul", "Aug", "Sep", "Oct"),
      drop = FALSE
    ) +
    theme(
      axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
      axis.title.x = element_blank()
    ) + 
    ylab("Relative Abundance \n (% of total community)") 


# Plot toxin over time
toxinplot <- 
  ggplot(data = erie_oligo_df, aes(x = Date, y = ParMC, group = Station)) +
    facet_grid(~Station) +
    geom_line(color = "black") +
    geom_point() +
  scale_x_discrete(
      breaks = c("7/8", "8/4", "9/2", "10/6"),
      labels = c("Jul", "Aug", "Sep", "Oct"),
      drop = FALSE
    ) +
  ylab("Particulate Microcystin-LR \n (ug/L)") + 
  theme(
      axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
      axis.title.x = element_blank(),
      plot.margin = unit(c(1, 0, 0, 0), "cm")
  )
   

```

```{r erie-oligo-toxin-plot, fig.width=10, fig.height=6}

# grab grobs for PC plots
toxinGrob <- ggplotGrob(toxinplot)
oligoGrob <- ggplotGrob(oligoplot)

toxinGrobWider <- gtable_add_cols(toxinGrob, widths = oligoGrob$widths[11])
toxinGrobWider <- gtable_add_cols(toxinGrobWider, widths = oligoGrob$widths[11])

# rbind the two plots together
plot <- rbind(oligoGrob, toxinGrobWider, size = "first")


erie_plot <- 
  ggdraw() +
    draw_plot(plot, x = 0.02, y = 0.02, width = 0.96, height = 0.94) + 
    draw_plot_label(c("A", "B"), c(0, 0), c(1, .52), size = 20) 

ggsave("../plots/raw-plots/erie-oligo-toxin-plot.pdf", plot = erie_plot, width = 10, height = 6)  

erie_plot

```

# Statistics

Median Chla 
```{r}

erie_oligo_df %>%
  group_by(Station) %>%
  summarise(median(Chla)) %>%
  pander()
```

Median TP levels
```{r}
erie_oligo_df %>%
  group_by(Station) %>%
  summarise(median(TP)) %>%
  pander()

```

## CTG, CCG, and toxicity 

Filter out dates with no microcystis 
```{r}
# Ratio of CTG to CCG
ratio <- 
  oligos_scale %>%
  filter((CTG + CCG + CTT) > 0) 
```

Look at the median ratio of CTG to CCG in July and August
```{r}
ratio %>%
  filter(Month %in% c( "July", "August")) %>%
  summarise(median(CTG_CCG_ratio)) %>%
  pander()
```

Look at the median ratio of CTG to CCG in September and October
```{r}

ratio %>%
  filter(Month %in% c("September", "October")) %>%
  summarise(median(CTG_CCG_ratio)) %>%
  pander()


```

Is there a significant correlation between CTG relative abundance and ParMC?
```{r}

CTG <-
  erie_oligo_df %>%
  filter(Oligotype == "CTG")

# Spearman's correlation
cor.test(CTG$count, CTG$ParMC, method = "spearman")

CTG_plot <- ggplot(CTG, aes(x = count, y = ParMC)) + 
  geom_point() + 
  ggtitle("CTG abundance vs Microcystin-LR toxin") + 
  xlab("CTG relative abundance") + 
  ylab("Microcystin-LR concentration \n (ug/L)")

ggsave("../plots/raw-plots/CTG_parmc.pdf", plot = CTG_plot, width = 7, height = 5) 

CTG_plot
```



## CTT and nutrient gradients
```{r}

# Mean relative abundance
CTT <- 
  erie_oligo_df %>%
    group_by(Shore) %>%
    filter(!is.na(count)) %>%
    filter(Oligotype == "CTT") 

CTT %>%
  summarise(median(count), min(count), max(count)) %>%
  pander()

```

Distribution of nearshore and offshore abundance of CTT
```{r CTT-distribution, fig.width=7, fig.height=5}

CTT_plot <- 
  ggplot(CTT, aes(x = count, group = Shore, fill = Shore)) +
    geom_density(alpha = .6, position = "identity") +
    scale_x_log10() +
    xlab("") +
    ggtitle("CTT relative abundance")

ggsave("../plots/raw-plots/CTT-distribution.pdf", plot = CTT_plot, width = 7, height = 5) 

CTT_plot
```

We will not calculate a statistic with a wilcoxon or permutational test,
because our data violates the assumption of independence. 


# Supplementary figure 2

Here we read in the RaxML tree for the cultures and calculate their 
patristic distances
```{r}

# Read in best raxml tree
rax_tree <- read.tree("data//RAxML_bestTree.bs100_mlst")

# Calculate patristic distances
patristic <- cophenetic(rax_tree)

# Remove the outgroup (NIES483) pulled from genbank
pat_dist <- as.dist(patristic[rownames(patristic) != "NIES483", colnames(patristic) != "NIES483"])

```

Here we calculate the hamming distance for the culture 16S-based 
oligotype sequence variants
```{r}

# Filter to just culture samples
culture_oligos <- 
  mc_oligos %>%
    mutate(sampID = substr(samples, 2, 10)) %>%
    filter(sampID %in% labels(pat_dist)) 


# Most cultures have reads from multiple oligotype sequences.
# This can be caused by sequencing errors, contamination during library prep, 
# or contamination in the cultures.
# This statement finds the consensus oligotype sequence for each culture i.e. the one with the maximum number of reads
cons_oligo <- apply(culture_oligos[,2:6], MARGIN = 1, function(x) {names(which.max(x))} )

# What percentage of the reads is the consensus sequence?
apply(culture_oligos[,2:6], MARGIN = 1, function(x) {max(x)/sum(x)} )
```
With one exception (76.9%), the consensus sequence constitutes 94-100% of the reads.

```{r}
# Add consensus oligotype to data frame
culture_oligos$ConsOligo <- cons_oligo

# Order the samples the same as they are ordered in the tree
target <- labels(pat_dist)
culture_oligos_sorted <- culture_oligos[match(target, culture_oligos$sampID), ]


# Calculate hamming distance
ham_dist <- adist(culture_oligos_sorted$ConsOligo)
rownames(ham_dist) <- culture_oligos_sorted$sampID
colnames(ham_dist) <- culture_oligos_sorted$sampID
ham_dist <- as.dist(ham_dist)

# Sanity check
all.equal(labels(ham_dist), labels(pat_dist))
```

Plot hamming distance vs patristic distance
```{r hamdist-vs-patristic-plot, fig.height=5, fig.width = 7}


df <- data.frame(patDist = as.vector(pat_dist), hamDist = as.vector(ham_dist))

df$hamDist <- as.factor(df$hamDist)

hamdist_plot <-
  ggplot(df, aes(x = hamDist, y = patDist, color = hamDist)) +
    geom_boxplot() + 
    ylab("Patristic Distance (RaxML tree)") +
    xlab("16S V4 hamming distance") +
    theme(legend.position = "none")

ggsave("../plots/raw-plots/hamdist-vs-patristic-plot.pdf", plot = hamdist_plot, width = 7, height = 5) 

hamdist_plot

```

```{r}
sessionInfo()

```

